### ISSUE TITLE
Weird nullref in ValueTuple.GetHashCode() and HashCode.Combine()

### ISSUE THREAD

**Issue by smoogipoo** [mono/mono#18973]
## Describing the issue

We've recently encountered a weird case where our app is freezing randomly. Upon further investigation with `MONO_GC_DEBUG=nursery-canaries,no-managed-allocator,check-remset-consistency`, I received the following callstack:

```
Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at System.ValueTuple`2[T1,T2].GetHashCode () <0x105aef2b0 + 0x00058> in <173a6557e95b49bebb395972201ed8fc#058e0622278f132a63191f97a5286050>:0
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_gsharedvt_constrained_call(intptr,intptr,intptr,intptr,intptr)
  at System.Collections.Generic.GenericEqualityComparer`1[T].GetHashCode (T obj) <0x104c789f0 + 0x00153> in <173a6557e95b49bebb395972201ed8fc#058e0622278f132a63191f97a5286050>:0
  at System.Collections.Concurrent.ConcurrentDictionary`2[TKey,TValue].TryGetValue (TKey key, TValue& value) <0x105aea6a8 + 0x0005b> in <173a6557e95b49bebb395972201ed8fc#058e0622278f132a63191f97a5286050>:0
  at osu.Framework.IO.Stores.FontStore.Get (System.String fontName, System.Char character) <0x10599f560 + 0x000ff> in <9c6cce5cde374c2bb191703496db58da#058e0622278f132a63191f97a5286050>:0
  at osu.Framework.Text.TextBuilder.getTexturedGlyph (Syst<…>
```

This references the following line:

https://github.com/ppy/osu-framework/blob/4d57525d82b0e55a5eae97e71a87004a6fa76f89/osu.Framework/IO/Stores/FontStore.cs#L137

At first I thought this was an issue with `ConcurrentDictionary` itself, since there was a period of time that structs could not be used as the key type for `Dictionary`.
Upon changing it to a `Dictionary` with appropriate `lock`s, the exception changed to the following:

```
System.NullReferenceException: Object reference not set to an instance of an object
  at System.ValueTuple`2[T1,T2].GetHashCode () <0x104295f00 + 0x00058> in <173a6557e95b49bebb395972201ed8fc#20ee7698523763452790178fddd3c870>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].FindEntry (TKey key) <0x10436bfc8 + 0x001a7> in <173a6557e95b49bebb395972201ed8fc#20ee7698523763452790178fddd3c870>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].TryGetValue (TKey key, TValue& value) <0x10434256c + 0x0004b> in <173a6557e95b49bebb395972201ed8fc#20ee7698523763452790178fddd3c870>:0 
  at osu.Framework.IO.Stores.FontStore.Get (System.String fontName, System.Char character) <0x103ff77d0 + 0x00107> in <c2c4cf2ad80a4615acd64e0011034834#20ee7698523763452790178fddd3c870>:0 
  at osu.Framework.Text.TextBuilder.getTexturedGlyph (System.Char character) <0x10410abd8 + 0x0004b> in <c2c4cf2ad80a4615acd64e0011034834#20ee7698523763452790178fddd3c870>:0 
  at osu.Framework.Text.TextB<…>
```

Finally, as a last resort I tried to manually create the hash code via: `var key = System.HashCode.Combine(fontName, character);`, which resulted in the following exception:

```
System.NullReferenceException: Object reference not set to an instance of an object
  at System.HashCode.Combine[T1,T2] (T1 value1, T2 value2) <0x10587a188 + 0x0004c> in <173a6557e95b49bebb395972201ed8fc#507cecd8c96df3a39850065ab75e9045>:0 
  at osu.Framework.IO.Stores.FontStore.Get (System.String fontName, System.Char character) <0x1055339b0 + 0x000af> in <936905cb8bb644ab987b84a43c268f7e#507cecd8c96df3a39850065ab75e9045>:0 
  at osu.Framework.Text.TextBuilder.getTexturedGlyph (System.Char character) <0x105646378 + 0x0004b> in <936905cb8bb644ab987b84a43c268f7e#507cecd8c96df3a39850065ab75e9045>:0 
  at osu.Framework.Text.TextBuilder.tryCreateGlyph (System.Char character, osu.Framework.Text.TextBuilderGlyph& glyph) <0x1056461d8 + 0x0003f> in <936905cb8bb644ab987b84a43c268f7e#507cecd8c96df3a39850065ab75e9045>:0 
  at osu.Framework.Text.TextBuilder.AddCharacter (System.Char character) <0x10564591c + 0x0005f> in <936905cb8bb644ab987b84a43c268f7e#507cecd8c96df3a39850065ab75e9045>:0 
```

At this point it looks like something internal to `HashCode` is failing, though it does seem like there's the appropriate null checks. I've tried to ensure that the string provided isn't null (essentially: `HashCode.Combine(fontName ?? string.Empty, character)`), but the exception doesn't change.

This occurs fairly randomly, so at this point I'm a little lost as to how I should be debugging this further. Nursery canaries shows no other issues.

## Repository

The repository in question can be found here: https://github.com/ppy/osu-framework

## Platforms

iOS

**Version Used**:

```
Mono Framework MDK
Runtime:
	Mono 6.8.0.104 (2019-10/1cca0cfebfc) (64-bit)
	Package version: 608000104

Xamarin.iOS
Version: 13.14.1.24 (Visual Studio Community)
Hash: fa00d7a6
Branch: d16-5
Build date: 2020-01-28 19:38:52-0500
```

**Comment by smoogipoo** [comment]
Furthermore, I've managed to capture a trace at the point this occurs (keep in mind the app freezes indefinitely, doesn't exit immediately).

https://drive.google.com/open?id=1TqtiHcB5Fs0C2_0HcOAKw5XZLink_z4I

Run 1 in particular, shows the nullref that I've reached above:

![image](https://user-images.githubusercontent.com/1329837/75014047-d3b0e780-54c8-11ea-876d-f57db38b76c5.png)

Although this looks impossible...

**Comment by marek-safar** [comment]
@BrzVlad could this be GC issue?

**Comment by smoogipoo** [comment]
I've split out the code into a small test project based on an empty Xamarin.iOS project, and couldn't replicate the issue. The memory corruption is likely happening before this dictionary is touched, or perhaps it's some sort of JIT issue?

Are there any more flags you'd recommend enabling to give more diagnostics for this issue? Managed memory dumps/etc that could be performed on iOS?

**Comment by pr0vieh** [comment]
maybe related to #18572

**Comment by BrzVlad** [comment]
The crash doesn't really look GC related. Do you encounter this crash without the debug options ? You can greatly impact the GC behavior and the reduce the amount of copied objects by doing something like MONO_GC_PARAMS=nursery-size=64M,evacuation-threshold=0 and MONO_GC_DEBUG=disable-minor. Would this affect the crash in any way ?

Given this looks more like a codegen problem, maybe you can try to enable/disable llvm.

Does it always hang ? The crash dump mechanism seems to hang in g_async_safe_printf. @lewurm @alexischr any ideas why we wouldn't be async safe ?

I could really use the assembly instruction where this crashes. How reliable is the reproduction test ? Could I try it locally ?

**Comment by peppy** [comment]
@BrzVlad the reproduction for us is quite consistent. 

I've included steps but please let me know if anything is unclear

- Build project from source (debug and release can both reproduce) at https://github.com/ppy/osu or install via testflight if that is more convenient (https://osu.ppy.sh/home/testflight)
- Download level via safari https://www.icloud.com/iclouddrive/0hXaFVrWZjYEN9L80cv4uwiiQ#961941_suzumu_-_Tokyo_Teddy_Bear
- From files, find the downloaded .osz file and Share -> Copy to osu!
- Enter song select by tapping the large pink logo a few times
- Select the hardest diffculty "Lost"
- Tap the "mods" button near the bottom of the screen
- Enable autoplay mod (with the cogs) and press "osu!" logo bottom right to start gameplay
- Seek forward in gameplay using the seek bar at the bottom of the screen
- The game may crash about 25%~50% in. If it finishes gameplay without a crash, running a second or third time should increase the chances of a crash. You can easily re-run from the results screen by pressing "back" (bottom-left) then "osu!" logo (bottom-right). It will retain the autoplay mode.

I've made a video showing these steps reproducing the crash: https://youtu.be/d28Tgnv24SA

Please let me know if anything is unclear.

**Comment by BrzVlad** [comment]
@peppy Are C# objects used outside of C# code ? Passed down to native code through pinvokes ?

**Comment by peppy** [comment]
As an update, I was mistaken. We cannot repro on debug or with LLVM disabled. we’ll loop back with more info on pinvokes and hopefully a simpler repro case.

**Comment by BrzVlad** [comment]
I could really use a smaller repro. Currently I'm not able to investigate since my iphone 6 cannot take on this large application and I'm getting crashes that I can't debug, and I'm not sure if they are related to bug in code or excessive memory usage.

**Comment by smoogipoo** [comment]
Apologies for the wait. I've been working on a thin reproduction project, which is completely automated.

Steps:
1. Clone https://github.com/smoogipoo/osu/tree/ios-crash-debug
2. Open osu.sln
3. Deploy osu.Game.Tests.iOS to device in `Release` mode.
4. Upon starting up, it will start the reproduction scenario automatically, involving a bunch of circles appearing on the screen.

When the app hangs (things stop moving), that indicates a successful reproduction.

On release config, the project is set up to use LLVM by default. You can confirm that it doesn't crash by disabling this.

Here's an example of how it should look in a successful reproduction: https://www.youtube.com/watch?v=dlmHGXoRs60

Note that it may take up to a minute to reproduce.

**Comment by lambdageek** [comment]
Is this an instance of https://github.com/mono/mono/pull/17227 - is the Dictionary keyed on a reference type that overrides GetHashCode and the value is a valuetype?

**Comment by smoogipoo** [comment]
Currently it's a `ValueTuple`, so I assume no unless it's somehow being boxed internally. As mentioned in the OP I've also replicated using just a simple `Dictionary` and afaik that does support value keys (as I am not sure of the exact characteristics of `ConcurrentDictionary`...).

I have also tested via an explicit `struct` and an explicit `class`, and both have experienced the same failure.

It's likely that whatever's going wrong is happening before this dictionary is accessed, and it just so happens that this is where the issue shows up.
